ytypes/node_cache.go

```diff
@@ -0,0 +1,274 @@
+package ytypes
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	json "github.com/goccy/go-json"
+	"github.com/openconfig/ygot/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// NodeCacheOpt is an option that can be usd with calls such as `SetNode` and `GetNode`.
+//
+// The `node cache` potentially provides fast-paths for config tree traversals and offers
+// noticeable performance boosts on a busy server that calls functions such as `SetNode`
+// and `GetNode` frequently.
+//
+// Passing in the reference of the `node cache` could prevent making the `ytypes` package
+// stateful. The applications that use the `ytypes` package maintain the `node cache`.
+type NodeCacheOpt struct {
+	NodeCache *NodeCache
+}
+
+// IsGetNodeOpt implements the GetNodeOpt interface.
+func (*NodeCacheOpt) IsGetNodeOpt() {}
+
+// IsSetNodeOpt implements the SetNodeOpt interface.
+func (*NodeCacheOpt) IsSetNodeOpt() {}
+
+// IsUnmarshalOpt marks IgnoreExtraFields as a valid UnmarshalOpt.
+func (*NodeCacheOpt) IsUnmarshalOpt() {}
+
+// IsGetOrCreateNodeOpt implements the GetOrCreateNodeOpt interface.
+func (*NodeCacheOpt) IsGetOrCreateNodeOpt() {}
+
+// cachedNodeInfo is used to provide shortcuts to making operations
+// to the nodes without having to traverse the config tree for every operation.
+type cachedNodeInfo struct {
+	parent interface{}
+	root   interface{}
+	nodes  []*TreeNode
+}
+
+// NodeCache is a thread-safe struct that's used for providing fast-paths for config tree traversals.
+type NodeCache struct {
+	store map[string]*cachedNodeInfo
+	mu    *sync.RWMutex
+}
+
+// NewNodeCache returns the pointer of a new `NodeCache` instance.
+func NewNodeCache() *NodeCache {
+	return &NodeCache{
+		store: map[string]*cachedNodeInfo{},
+		mu:    &sync.RWMutex{},
+	}
+}
+
+// Reset resets the cache. The cache will be repopulated after the reset.
+func (c *NodeCache) Reset() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.store = map[string]*cachedNodeInfo{}
+}
+
+// setNodeCache uses the cached information to set the node instead of traversalling the config tree.
+// This improves runtime performance of the library.
+func (c *NodeCache) set(path *gpb.Path, val interface{}, opts ...SetNodeOpt) (setComplete bool, err error) {
+	switch val.(type) {
+	case *gpb.TypedValue:
+	default:
+		// Only TypedValue is supported by the node cache for now.
+		return
```

Comment:
Thanks for the comment. I think this type switch is used to check the type of `val` and see if it's supported by the node cache. Maybe we don't want to return an error since that will turn into a `SetNode` error and block the `Set`. 

The node cache doesn't support all types that the ordinary `SetNode` supports so the intention is to let the `SetNode` continue (which will still modify the node, without the performance boost) if the `val` is not in supported type. Hope that makes sense.